presto-main/src/main/java/com/facebook/presto/sql/relational/optimizer/ExpressionOptimizer.java

```diff
@@ -62,26 +62,32 @@ public ExpressionOptimizer(FunctionRegistry registry, TypeManager typeManager, S
 
     public RowExpression optimize(RowExpression expression)
     {
-        return expression.accept(new Visitor(), null);
+        return expression.accept(new Visitor(), ExceptionBehavior.PROPAGATE_EXCEPTION);
+    }
+
+    private enum ExceptionBehavior
+    {
+        KEEP_UNEVALUATED_EXPRESSION_AND_CONTINUE,
```

Comment:
I don't think we need to parameterize the behavior like this. It's always safe to return the expression partially un-optimized in case of error and let the runtime deal with it.